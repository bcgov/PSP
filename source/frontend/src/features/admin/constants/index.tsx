import { ColumnWithProps } from 'components/Table';
import { IOrganization, IOrganizationRecord } from 'interfaces';
import * as React from 'react';
import { Link } from 'react-router-dom';
import { CellProps } from 'react-table';

export const columnDefinitions: ColumnWithProps<IOrganizationRecord>[] = [
  {
    Header: 'Organization name',
    accessor: 'name',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Short name',
    accessor: 'code',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Description',
    accessor: 'description',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Parent Organization',
    accessor: 'parent',
    align: 'left',
    clickable: true,
    Cell: (props: CellProps<IOrganization>) => {
      return (
        <Link to={`/admin/organization/${props.row.original.parentId}`}>
          {props.row.original.parent}
        </Link>
      );
    },
  },
];
