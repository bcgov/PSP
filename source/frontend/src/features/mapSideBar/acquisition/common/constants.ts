export const cannotEditMessage =
  'This section of the file is in a non-editable state. Contact a System administrator to change it to an editable state';
