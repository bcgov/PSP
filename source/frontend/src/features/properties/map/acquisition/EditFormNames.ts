export enum EditFormNames {
  acquisitionSummary = 'acquisitionSummary',
  acquisitionChecklist = 'acquisitionChecklist',
  propertyDetails = 'propertyDetails',
  propertySelector = 'propertySelector',
  takes = 'takes',
}
