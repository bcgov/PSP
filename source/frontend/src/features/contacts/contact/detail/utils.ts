import { ContactInfoField } from 'features/contacts/interfaces';
import { Dictionary } from 'interfaces/Dictionary';
import { IContactAddress, IContactMethod } from 'interfaces/IContact';

import { AddressTypes } from './../../../../constants/addressTypes';

// the order of this array corresponds to the expected display order
const addressSortOrder = [AddressTypes.Mailing, AddressTypes.Residential, AddressTypes.Billing];

export const sortAddresses = (a1: IContactAddress, a2: IContactAddress) => {
  if (a2.addressType.id === a1.addressType.id) {
    return 0;
  }
  const a2Index = addressSortOrder.indexOf(a2.addressType.id as AddressTypes);
  if (a2Index !== -1 && a2Index < addressSortOrder.indexOf(a1.addressType.id as AddressTypes)) {
    return 1;
  }

  return -1;
};

export function getContactInfo(
  validTypes: Dictionary<string>,
  contactMethods?: IContactMethod[],
): ContactInfoField[] {
  if (contactMethods === undefined) {
    return [];
  }
  // Get only the valid types
  let filteredFields = contactMethods.reduce(
    (accumulator: ContactInfoField[], method: IContactMethod) => {
      if (Object.keys(validTypes).includes(method.contactMethodType.id)) {
        accumulator.push({
          info: method.value,
          label: validTypes[method.contactMethodType.id],
        });
      }
      return accumulator;
    },
    [],
  );

  // Sort according to the dictionary order
  return filteredFields.sort((a, b) => {
    return Object.values(validTypes).indexOf(a.label) - Object.values(validTypes).indexOf(b.label);
  });
}

export const fakeAddresses = [
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'BILLING',
      description: 'Billing address',
      isDisabled: false,
    },
    streetAddress1: 'Billing address',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6A 5G7',
  },
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'MAILING',
      description: 'Mailing address',
      isDisabled: false,
    },
    streetAddress1: 'Mailing address',
    streetAddress2: 'Living in a van',
    streetAddress3: 'Down by the River',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6Z 5G7',
  },
  {
    id: 5,
    rowVersion: 1,
    addressType: {
      id: 'RESIDNT',
      description: 'Property address',
      isDisabled: false,
    },
    streetAddress1: 'Property address',
    streetAddress2: '',
    streetAddress3: '',
    municipality: 'vic',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V9A 0H6',
  },
];
