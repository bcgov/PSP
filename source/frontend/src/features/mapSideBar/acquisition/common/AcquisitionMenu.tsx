import cx from 'classnames';
import { Col, Row } from 'react-bootstrap';
import { FaCaretRight } from 'react-icons/fa';
import styled from 'styled-components';

import { EditPropertiesIcon } from '@/components/common/buttons/EditPropertiesButton';
import { LinkButton } from '@/components/common/buttons/LinkButton';
import EditButton from '@/components/common/EditButton';
import TooltipIcon from '@/components/common/TooltipIcon';
import { Claims, Roles } from '@/constants/index';
import { useKeycloakWrapper } from '@/hooks/useKeycloakWrapper';
import { ApiGen_Concepts_AcquisitionFile } from '@/models/api/generated/ApiGen_Concepts_AcquisitionFile';

import AcquisitionFileStatusUpdateSolver from '../tabs/fileDetails/detail/AcquisitionFileStatusUpdateSolver';
import { cannotEditMessage } from './constants';
import GenerateFormContainer from './GenerateForm/GenerateFormContainer';
import GenerateFormView from './GenerateForm/GenerateFormView';

export interface IAcquisitionMenuProps {
  acquisitionFile: ApiGen_Concepts_AcquisitionFile;
  items: string[];
  selectedIndex: number;
  onChange: (index: number) => void;
  onShowPropertySelector: () => void;
}

const AcquisitionMenu: React.FunctionComponent<
  React.PropsWithChildren<IAcquisitionMenuProps>
> = props => {
  const { hasClaim, hasRole } = useKeycloakWrapper();
  const handleClick = (index: number) => {
    props.onChange(index);
  };
  const statusSolver = new AcquisitionFileStatusUpdateSolver(
    props.acquisitionFile.fileStatusTypeCode,
  );
  const canEditDetails = () => {
    if (hasRole(Roles.SYSTEM_ADMINISTRATOR) || statusSolver.canEditProperties()) {
      return true;
    }
    return false;
  };
  return (
    <>
      <StyledMenuWrapper>
        {props.items.map((label: string, index: number) => {
          const activeIndex = props.selectedIndex === index;
          if (index === 0) {
            return (
              <StyledRow
                key={`menu-item-row-${index}`}
                data-testid={`menu-item-row-${index}`}
                className={cx('no-gutters', { selected: props.selectedIndex === index })}
              >
                {activeIndex ? (
                  <StyledMenuCol>
                    <span title="File Details">{label}</span>
                  </StyledMenuCol>
                ) : (
                  <Col>
                    <LinkButton title="File Details" onClick={() => handleClick(index)}>
                      {label}
                    </LinkButton>
                  </Col>
                )}
                <StyledMenuHeaderWrapper>
                  <StyledMenuHeader>Properties</StyledMenuHeader>
                  {hasClaim(Claims.ACQUISITION_EDIT) && canEditDetails() && (
                    <EditButton
                      title="Change properties"
                      icon={<EditPropertiesIcon />}
                      onClick={props.onShowPropertySelector}
                    />
                  )}
                  {hasClaim(Claims.ACQUISITION_EDIT) && !canEditDetails() && (
                    <TooltipIcon
                      toolTipId={`${props?.acquisitionFile?.id || 0}-summary-cannot-edit-tooltip`}
                      toolTip={cannotEditMessage}
                    />
                  )}
                </StyledMenuHeaderWrapper>
              </StyledRow>
            );
          } else {
            return (
              <StyledRow
                key={`menu-item-row-${index}`}
                data-testid={`menu-item-row-${index}`}
                className={cx('no-gutters', { selected: props.selectedIndex === index })}
                onClick={() => (props.selectedIndex !== index ? handleClick(index) : '')}
              >
                <Col xs="1">{props.selectedIndex === index && <FaCaretRight />}</Col>
                <Col xs="auto" className="pr-2">
                  <StyledIconWrapper className={cx({ selected: props.selectedIndex === index })}>
                    {index}
                  </StyledIconWrapper>
                </Col>
                {activeIndex ? (
                  <StyledMenuCol>
                    <span title="View">{label}</span>
                  </StyledMenuCol>
                ) : (
                  <Col>
                    <LinkButton title="View">{label}</LinkButton>
                  </Col>
                )}
              </StyledRow>
            );
          }
        })}
      </StyledMenuWrapper>
      {props.acquisitionFile?.id && (
        <GenerateFormContainer
          acquisitionFileId={props.acquisitionFile.id}
          View={GenerateFormView}
        />
      )}
    </>
  );
};

export default AcquisitionMenu;

const StyledMenuWrapper = styled.div`
  text-align: left;
  padding: 0px;
  margin: 0px;
  width: 100%;
  color: ${props => props.theme.css.linkColor};
`;

const StyledMenuCol = styled(Col)`
  min-height: 2.5rem;
  line-height: 3rem;
`;

const StyledRow = styled(Row)`
  &.selected {
    font-weight: bold;
    cursor: default;
  }

  font-size: 1.4rem;
  font-weight: normal;
  cursor: pointer;
  padding-bottom: 0.5rem;

  div.Button__value {
    font-size: 1.4rem;
  }
`;

const StyledIconWrapper = styled.div`
  &.selected {
    background-color: ${props => props.theme.bcTokens.themeGold100};
  }

  background-color: ${props => props.theme.css.numberBackgroundColor};
  font-size: 1.5rem;
  border-radius: 50%;
  opacity: 0.8;
  width: 2.5rem;
  height: 2.5rem;
  padding: 1rem;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const StyledMenuHeaderWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  width: 100%;
  border-bottom: 1px solid ${props => props.theme.css.borderOutlineColor};
`;

const StyledMenuHeader = styled.span`
  font-weight: bold;
  font-size: 1.6rem;
  color: ${props => props.theme.bcTokens.iconsColorSecondary};
  line-height: 2.2rem;
`;
