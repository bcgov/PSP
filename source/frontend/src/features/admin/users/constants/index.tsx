import { ReactComponent as Active } from 'assets/images/active.svg';
import { ReactComponent as Inactive } from 'assets/images/inactive.svg';
import { ColumnWithProps } from 'components/Table';
import { DateTimeCell } from 'components/Table/DateCell';
import { Link } from 'react-router-dom';
import { CellProps } from 'react-table';
import { stringToFragment } from 'utils';

import { RowActions } from '../components/RowActions';
import { FormUser } from '../models';

export const getUserColumns = (refresh: () => void): ColumnWithProps<FormUser>[] => [
  {
    Header: 'Active',
    accessor: 'isDisabled',
    align: 'left',
    clickable: true,
    sortable: true,
    width: 60,
    Cell: (props: CellProps<FormUser>) =>
      props.row.original.isDisabled ? <Inactive /> : <Active />,
  },
  {
    Header: 'IDIR/BCeID',
    accessor: 'businessIdentifierValue',
    align: 'left',
    minWidth: 150,
    Cell: (props: CellProps<FormUser>) => {
      return (
        <Link to={`/admin/user/${props.row.original.id}`}>
          {props.row.original.businessIdentifierValue}
        </Link>
      );
    },
    sortable: true,
  },
  {
    Header: 'First name',
    accessor: 'firstName',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Last name',
    accessor: 'surname',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Email',
    accessor: 'email',
    align: 'left',
    clickable: true,
    sortable: true,
    minWidth: 200,
  },
  {
    Header: 'Position',
    accessor: 'position',
    align: 'left',
    clickable: true,
    sortable: true,
  },

  {
    Header: 'Roles',
    accessor: 'roles',
    align: 'left',
    clickable: true,
    minWidth: 200,
    Cell: (props: CellProps<FormUser>) => {
      const rolesString = props.row?.original?.roles?.map(userRole => userRole?.name);
      return stringToFragment(rolesString?.join(', '));
    },
  },
  {
    Header: 'MoTI region(s)',
    accessor: 'regions',
    align: 'left',
    clickable: true,
    minWidth: 200,
    Cell: (props: CellProps<FormUser>) => {
      const regionsString = props.row?.original?.regions?.map(
        regionUser => regionUser?.description,
      );
      return stringToFragment(regionsString?.join(', '));
    },
  },
  {
    Header: 'Last login',
    accessor: 'lastLogin',
    align: 'left',
    clickable: true,
    minWidth: 100,
    Cell: DateTimeCell,
  },
  {
    Header: ' ',
    Cell: (props: CellProps<FormUser>) => {
      return <RowActions {...props} refresh={refresh} />;
    },
    width: 75,
  },
];
