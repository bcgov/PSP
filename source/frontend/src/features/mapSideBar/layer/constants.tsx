import { Feature } from 'geojson';
import { FaExternalLinkAlt } from 'react-icons/fa';

import { ApiGen_Concepts_Address } from '@/models/api/generated/ApiGen_Concepts_Address';
import { exists, formatApiAddress, getAreaUnit, pidFormatter } from '@/utils';

import { ContentConfig } from './LayerContent';

export const parcelLayerConfig: ContentConfig = {
  PID: { label: 'Parcel PID', display: (data: { [key: string]: any }) => pidFormatter(data?.PID) },
  PIN: { label: 'Parcel PIN', display: (data: { [key: string]: any }) => data.PIN },
  PLAN_NUMBER: {
    label: 'Plan number',
    display: (data: { [key: string]: any }) => data.PLAN_NUMBER,
  },
  LEGAL_DESCRIPTION: {
    label: 'Legal description',
    display: (data: { [key: string]: any }) => data.LEGAL_DESCRIPTION,
  },
  OWNER_TYPE: { label: 'Owner type', display: (data: { [key: string]: any }) => data.OWNER_TYPE },
  MUNICIPALITY: {
    label: 'Municipality',
    display: (data: { [key: string]: any }) => data.MUNICIPALITY,
  },
  REGIONAL_DISTRICT: {
    label: 'Regional district',
    display: (data: { [key: string]: any }) => data.REGIONAL_DISTRICT,
  },
  FEATURE_AREA_SQM: {
    label: 'Area',
    display: (data: { [key: string]: any }) => (
      <>
        {data.FEATURE_AREA_SQM} m<sup>2</sup>
      </>
    ),
  },
  PARCEL_CLASS: {
    label: 'Parcel class',
    display: (data: { [key: string]: any }) => data.PARCEL_CLASS,
  },
  SURVEY_DESIGNATION_1: {
    label: 'Survey designation one',
    display: (data: { [key: string]: any }) => data.SURVEY_DESIGNATION_1,
  },
  SURVEY_DESIGNATION_2: {
    label: 'Survey designation two',
    display: (data: { [key: string]: any }) => data.SURVEY_DESIGNATION_2,
  },
  SURVEY_DESIGNATION_3: {
    label: 'Survey designation three',
    display: (data: { [key: string]: any }) => data.SURVEY_DESIGNATION_3,
  },
};

// Utility function that dynamically creates a ContentConfig for a feature.
// Transforms "THE_PARAM" to "The param"
export const getDynamicFeatureConfig: (feature: Feature) => ContentConfig = (feature: Feature) => {
  // For debug enable this
  /*
  config['Feature Type'] = {
    label: 'Feature Type',
    display: (data: { [key: string]: any }) => feature.id,
  };
  */
  const config: ContentConfig = {};
  Object.keys(feature.properties).forEach(key => {
    config[key] = {
      label: key.charAt(0).toUpperCase() + key.slice(1).toLowerCase().replaceAll('_', ' '),
      display: (data: { [key: string]: any }) => data[key],
    };
  });
  // This is not being mapped in the typed feature data
  if (exists(config['SE_ANNO_CAD_DATA'])) {
    delete config['SE_ANNO_CAD_DATA'];
  }
  return config;
};

export const municipalityLayerConfig: ContentConfig = {
  MUNICIPALITY_NAME: {
    label: 'Municipality Name',
    display: (data: { [key: string]: any }) =>
      `${data.ADMIN_AREA_GROUP_NAME} (${data.CHANGE_REQUESTED_ORG})`,
  },
};

export const pimsLayerConfig: ContentConfig = {
  PROPERTY_ID: {
    label: 'Hyperlink',
    display: (data: { [key: string]: any }) => (
      <a
        href={`/mapview/sidebar/property/${data.PROPERTY_ID}/details`}
        target="_blank"
        rel="noreferrer"
      >
        View PIMS property&nbsp;
        <FaExternalLinkAlt />
      </a>
    ),
  },
  PID: {
    label: 'PID',
    display: (data: { [key: string]: any }) => pidFormatter(data.PID_PADDED),
  },
  PIN: {
    label: 'PIN',
    display: (data: { [key: string]: any }) => data.PIN,
  },
  UNIQUE_ID: {
    label: 'Plan number',
    display: (data: { [key: string]: any }) => data.SURVEY_PLAN_NUMBER,
  },
  IS_RETIRED: {
    label: 'Retired ?',
    display: (data: { [key: string]: any }) => (data.IS_RETIRED ? 'Yes' : 'No'),
  },
  LAND_LEGAL_DESCRIPTION: {
    label: 'PIMS legal description',
    display: (data: { [key: string]: any }) => data.LAND_LEGAL_DESCRIPTION,
  },

  PIMS_ADDRESS: {
    label: 'Pims Address',
    display: (data: { [key: string]: any }) =>
      formatApiAddress({
        streetAddress1: data.STREET_ADDRESS_1,
        streetAddress2: data.STREET_ADDRESS_2,
        streetAddress3: data.STREET_ADDRESS_3,
        postal: data.POSTAL_CODE,
        province: { code: 'BC' },
        municipality: data.MUNICIPALITY_NAME,
      } as ApiGen_Concepts_Address), // force the type to avoid having to list unused fields.
  },
  LAND_AREA: {
    label: 'PIMS land area',
    display: (data: { [key: string]: any }) =>
      data.LAND_AREA + ' ' + getAreaUnit(data.PROPERTY_AREA_UNIT_TYPE_CODE.toLowerCase()),
  },
};

export const highwayLayerConfig: ContentConfig = {
  PROVINCIAL_PUBLIC_HIGHWAY_ID: {
    label: 'Id',
    display: (data: { [key: string]: any }) => data.PROVINCIAL_PUBLIC_HIGHWAY_ID,
  },
  GLOBALID: { label: 'Global Id', display: (data: { [key: string]: any }) => data.GLOBALID },
  UNIQUE_ID: { label: 'Unique Id', display: (data: { [key: string]: any }) => data.UNIQUE_ID },
  HYPERLINK: {
    label: 'Hyperlink to plan documents',
    display: (data: { [key: string]: any }) => (
      <a href={`${data.HYPERLINK}`} target="_blank" rel="noreferrer">
        {decodeURI(data.HYPERLINK)}
      </a>
    ),
  },
  PLAN_ANNOTATION: {
    label: 'Plan annotation',
    display: (data: { [key: string]: any }) => data.PLAN_ANNOTATION,
  },
  MOTI_PLAN: { label: 'MOTI plan', display: (data: { [key: string]: any }) => data.MOTI_PLAN },
  VETTING_STATUS: {
    label: 'Vetting Status',
    display: (data: { [key: string]: any }) => data.VETTING_STATUS,
  },
  SHAPE_TYPE: { label: 'Shape type', display: (data: { [key: string]: any }) => data.SHAPE_TYPE },
  PENDING_CLASSIFICATION: {
    label: 'Pending classification',
    display: (data: { [key: string]: any }) => data.PENDING_CLASSIFICATION,
  },
  MOTI_FILE: { label: 'MOTI file', display: (data: { [key: string]: any }) => data.MOTI_FILE },
  GAZETTE_PUBLISHED_DATE: {
    label: 'Gazette published date',
    display: (data: { [key: string]: any }) => data.GAZETTE_PUBLISHED_DATE,
  },
  GAZETTE_PUBLISHED_ON: {
    label: 'Gazette published on',
    display: (data: { [key: string]: any }) => data.GAZETTE_PUBLISHED_ON,
  },
  PROV_PUBLIC_HIGHWAY_TYPE: {
    label: 'Prov public highway type',
    display: (data: { [key: string]: any }) => data.PROV_PUBLIC_HIGHWAY_TYPE,
  },
  ORDER_IN_COUNCIL: {
    label: 'Order in council',
    display: (data: { [key: string]: any }) => data.ORDER_IN_COUNCIL,
  },
  SHORT_LEGAL_DESCRIPTION: {
    label: 'Short legal description',
    display: (data: { [key: string]: any }) => data.SHORT_LEGAL_DESCRIPTION,
  },
  SURVEY_PLAN: {
    label: 'Survey plan',
    display: (data: { [key: string]: any }) => data.SURVEY_PLAN,
  },
  PARENT_PARCEL: {
    label: 'Parent parcel',
    display: (data: { [key: string]: any }) => data.PARENT_PARCEL,
  },
  PARCELMAPBC_COMMENT: {
    label: 'Parcelmapbc comment',
    display: (data: { [key: string]: any }) => data.PARCELMAPBC_COMMENT,
  },
  CREATE_TIME: {
    label: 'Create time',
    display: (data: { [key: string]: any }) => data.CREATE_TIME,
  },
  UPDATE_TIME: {
    label: 'Update time',
    display: (data: { [key: string]: any }) => data.UPDATE_TIME,
  },
  CREATE_USER: {
    label: 'Create user',
    display: (data: { [key: string]: any }) => data.CREATE_USER,
  },
  UPDATE_USER: {
    label: 'Update user',
    display: (data: { [key: string]: any }) => data.UPDATE_USER,
  },
  LTSA_CREATE_TIME: {
    label: 'Ltsa create time',
    display: (data: { [key: string]: any }) => data.LTSA_CREATE_TIME,
  },
  LTSA_UPDATE_TIME: {
    label: 'Ltsa update time',
    display: (data: { [key: string]: any }) => data.LTSA_UPDATE_TIME,
  },
  LTSA_CREATE_USER: {
    label: 'Ltsa create user',
    display: (data: { [key: string]: any }) => data.LTSA_CREATE_USER,
  },
  LTSA_UPDATE_USER: {
    label: 'Ltsa update user',
    display: (data: { [key: string]: any }) => data.LTSA_UPDATE_USER,
  },
  RELATED_GAZETTES: {
    label: 'Related gazettes',
    display: (data: { [key: string]: any }) => data.RELATED_GAZETTES,
  },
  RELATIVE_PATH: {
    label: 'Relative path',
    display: (data: { [key: string]: any }) => data.RELATIVE_PATH,
  },
};
