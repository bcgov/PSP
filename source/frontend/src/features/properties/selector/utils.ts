import { IMapProperty } from 'features/properties/selector/models';
import { compact } from 'lodash';
import { Api_PropertyFile } from 'models/api/PropertyFile';
import { formatApiAddress, pidFormatter } from 'utils';

import { Api_Geometry, Api_Property } from './../../../models/api/Property';

export enum NameSourceType {
  PID = 'PID',
  PIN = 'PIN',
  PLAN = 'Plan #',
  LOCATION = 'Location',
  NAME = 'Descriptive Name',
  NONE = 'n/a',
  ADDRESS = 'Address',
}

interface PropertyName {
  label: NameSourceType;
  value: string;
}

export const getPropertyName = (property: IMapProperty): PropertyName => {
  if (property.pid !== undefined && property.pid?.toString().length > 0 && property.pid !== '0') {
    return { label: NameSourceType.PID, value: pidFormatter(property.pid.toString()) };
  } else if (
    property.pin !== undefined &&
    property.pin?.toString()?.length > 0 &&
    property.pin !== '0'
  ) {
    return { label: NameSourceType.PIN, value: property.pin.toString() };
  } else if (property.planNumber !== undefined) {
    return { label: NameSourceType.PLAN, value: property.planNumber };
  } else if (property.latitude !== undefined && property.longitude !== undefined) {
    return {
      label: NameSourceType.LOCATION,
      value: compact([property.longitude?.toFixed(6), property.latitude?.toFixed(6)]).join(', '),
    };
  } else if (property.address !== undefined) {
    return {
      label: NameSourceType.ADDRESS,
      value: property.address,
    };
  }
  return { label: NameSourceType.NONE, value: '' };
};

export const getPrettyLatLng = (location?: Api_Geometry) =>
  compact([
    location?.coordinate?.x?.toFixed(6) ?? 0,
    location?.coordinate?.y?.toFixed(6) ?? 0,
  ]).join(', ');

export const getFilePropertyName = (
  fileProperty?: Api_PropertyFile,
  skipName: boolean = false,
): PropertyName => {
  if (fileProperty === undefined) {
    return { label: NameSourceType.NONE, value: '' };
  }

  if (
    fileProperty.propertyName !== undefined &&
    fileProperty.propertyName !== '' &&
    skipName === false
  ) {
    return { label: NameSourceType.NAME, value: fileProperty.propertyName };
  } else if (fileProperty.property !== undefined) {
    const property = fileProperty.property;
    return getApiPropertyName(property);
  }
  return { label: NameSourceType.NONE, value: '' };
};

export const getApiPropertyName = (property?: Api_Property): PropertyName => {
  if (property === undefined) {
    return { label: NameSourceType.NONE, value: '' };
  }

  let mapProperty: IMapProperty = {
    pin: property.pin?.toString(),
    pid: property.pid?.toString(),
    latitude: property.latitude,
    longitude: property.longitude,
    planNumber: property.planNumber,
    address: property.address !== undefined ? formatApiAddress(property.address) : undefined,
  };
  return getPropertyName(mapProperty);
};
