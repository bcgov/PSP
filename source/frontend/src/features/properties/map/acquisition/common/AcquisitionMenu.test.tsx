import { Claims } from 'constants/index';
import { render, RenderOptions, userEvent } from 'utils/test-utils';

import { EditFormNames } from '../EditFormNames';
import AcquisitionMenu, { IAcquisitionMenuProps } from './AcquisitionMenu';

// mock auth library
jest.mock('@react-keycloak/web');

const onChange = jest.fn();
const setContainerState = jest.fn();

const testData = ['one', 'two', 'three'];

describe('AcquisitionMenu component', () => {
  // render component under test
  const setup = (props: IAcquisitionMenuProps, renderOptions: RenderOptions = {}) => {
    const utils = render(
      <AcquisitionMenu
        selectedIndex={props.selectedIndex}
        items={props.items}
        onChange={props.onChange}
        setContainerState={props.setContainerState}
      />,
      {
        useMockAuthentication: true,
        claims: [Claims.ACQUISITION_EDIT],
        ...renderOptions,
      },
    );

    return { ...utils };
  };

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('matches snapshot', () => {
    const { asFragment } = setup({
      items: testData,
      selectedIndex: 0,
      onChange,
      setContainerState,
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the items ', () => {
    const { getByText } = setup({
      items: testData,
      selectedIndex: 0,
      onChange,
      setContainerState,
    });

    expect(getByText('one')).toBeVisible();
    expect(getByText('two')).toBeVisible();
    expect(getByText('three')).toBeVisible();
  });

  it('renders selected item with different style', () => {
    const { getByTestId } = setup({
      items: testData,
      selectedIndex: 1,
      onChange,
      setContainerState,
    });

    expect(getByTestId('menu-item-row-0')).not.toHaveClass('selected');
    expect(getByTestId('menu-item-row-1')).toHaveClass('selected');
    expect(getByTestId('menu-item-row-2')).not.toHaveClass('selected');
  });

  it('allows the selected item to be changed', () => {
    const { getByText } = setup({
      items: testData,
      selectedIndex: 1,
      onChange,
      setContainerState,
    });

    const lastItem = getByText('three');
    userEvent.click(lastItem);

    expect(onChange).toHaveBeenCalledWith(2);
  });

  it(`renders the edit button for users with property edit permissions`, () => {
    const { getByTitle } = setup(
      {
        items: testData,
        selectedIndex: 1,
        onChange,
        setContainerState,
      },
      { claims: [Claims.ACQUISITION_EDIT] },
    );

    const button = getByTitle('Change properties');
    expect(button).toBeVisible();

    userEvent.click(button);

    expect(setContainerState).toHaveBeenCalledWith({
      isEditing: true,
      activeEditForm: EditFormNames.propertySelector,
    });
  });

  it(`doesn't render the edit button for users without edit permissions`, () => {
    const { queryByTitle } = setup(
      {
        items: testData,
        selectedIndex: 1,
        onChange,
        setContainerState,
      },
      { claims: [Claims.ACQUISITION_VIEW] }, // no edit permissions, just view.
    );

    const button = queryByTitle('Change properties');
    expect(button).toBeNull();
  });
});
