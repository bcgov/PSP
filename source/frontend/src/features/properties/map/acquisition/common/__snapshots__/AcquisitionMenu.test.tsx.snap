// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AcquisitionMenu component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="container"
  />
  <div
    class="Toastify"
  />
  .c0 {
  text-align: left;
  padding: 0px;
  margin: 0px;
  width: 100%;
}

.c1 {
  font-weight: normal;
  cursor: pointer;
  padding-bottom: 0.5rem;
}

.c1.selected {
  font-weight: bold;
  cursor: default;
}

.c4 {
  font-size: 1.5rem;
  border-radius: 50%;
  width: 2.5rem;
  height: 2.5rem;
  padding: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  margin-top: 1rem;
  border-bottom: 1px solid;
}

.c3 {
  font-size: 1.4rem;
  line-height: 2.2rem;
}

<div
    class="c0"
  >
    <div
      class="c1 no-gutters selected row"
      data-testid="menu-item-row-0"
    >
      <div
        class="col-1"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 192 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
          />
        </svg>
      </div>
      <div
        class="col"
      >
        one
      </div>
      <div
        class="c2"
      >
        <span
          class="c3"
        >
          Properties
        </span>
        <button
          class="btn btn-link"
          title="Edit acquisition properties"
          type="button"
        >
          <svg
            fill="currentColor"
            height="2rem"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 576 512"
            width="2rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="c1 no-gutters row"
      data-testid="menu-item-row-1"
    >
      <div
        class="col-1"
      />
      <div
        class="pr-2 col-auto"
      >
        <div
          class="c4"
        >
          1
        </div>
      </div>
      <div
        class="col"
      >
        two
      </div>
    </div>
    <div
      class="c1 no-gutters row"
      data-testid="menu-item-row-2"
    >
      <div
        class="col-1"
      />
      <div
        class="pr-2 col-auto"
      >
        <div
          class="c4"
        >
          2
        </div>
      </div>
      <div
        class="col"
      >
        three
      </div>
    </div>
  </div>
</DocumentFragment>
`;
